src/tranquility/mini/visitor.lua

```diff
@@ -0,0 +1,273 @@
+-- TODO: checking for "" is still a bit ugly, but all is functional now
+require("tranquility.mini.grammar")
+local unpack = table.unpack or _G.unpack
+Visitor = {}
+
+function Visitor:visit(node)
+	local type = node[1]
+	local method = self[type]
+	if node[3] ~= nil then
+		local children = {}
+		for i = 3, #node do
+			children[#children + 1] = node[i]
+		end
+		for i, subnode in pairs(children) do
+			children[i] = self:visit(subnode)
+		end
+		return method(self, node, children)
+	else
+		return method(self, node, "")
+	end
+end
+
+function Visitor:root(_, children)
+	return children[1]
+end
+
+function Visitor:sequence(_, children)
+	local group, other_groups, other_seqs = unpack(children)
+	if other_groups ~= "" then
+		local elements = {
+			{ type = "element", value = { type = "polyrhythm", seqs = { group } }, modifiers = {} },
+		}
+		for _, item in pairs(other_groups) do
+			elements[#elements + 1] = {
+				type = "element",
+				value = {
+					type = "polyrhythm",
+					seqs = { item },
+				},
+				modifiers = {},
+			}
+		end
+		group = { type = "sequence", elements = elements }
+	end
+	if other_seqs ~= "" then
+		local n_seq = { group }
+		for _, item in pairs(other_seqs) do
+			n_seq[#n_seq + 1] = item
+		end
+		return { type = "random_sequence", elements = n_seq }
+	end
+	return group
+end
+
+-- need a test?
```

Comment:
I'm feeling good with the test you have. It tests the publicly facing functionality (ie. we end up with the correct pattern in the end). I'm fine thinking about exactly how you parsed the AST as an internal implementation detail. Do you agree?

Reply:
oh yes this is left there for some reason.
About the quality of the test, the targets are from vortex tests, and I also added a few, which is also based on the output of vortex.